### ISSUE TITLE
Visibility into commands being run?

### ISSUE THREAD

**Issue by lesteve** [scientific-python/spin#161]
I mentioned I was interested to look at spin in the last scikit-learn developer meeting. One feed-back I got was the worry that it adds another layer of tooling. There is some concern to end-up in a situation where there are too many layers. For example, `spin` that calls `make` that calls a shell script, that calls a Python script, etc ...

I looked a bit to see whether there was some built-in spin functionality that would somehow give some visibility into which commands gets run (kind of `make --dry-run` for example). This would allow to understand a bit more what gets run behind the scenes and potentially do it without spin if needed e.g. to put together a reproducer.

For now what I found is that you can add a verbose flag by command and it is the responsability of the command implementation to put whatever makes sense behind this verbose flag (or even without this verbose flag by the way). The meson commands seems to be quite verbose by default (which is a good thing in this case) and show which commands gets run. If we implement custom commands we will have to do it.

Did I get this roughly correctly?

**Comment by stefanv** [comment]
At least for the built-in commands in Meson, we always print what we do behind the scenes, unless it is problematic for the command. E.g., the `run` command cannot always print to stdout, because you may want to do `OUTPUT=$(spin run ...)`. But otherwise you should see `$ shell command` each time we execute something.

The purpose is exactly to address your concern: that there is no extra layer of complexity, simply a layer of convenience. You should be able to always just cut and paste commands from the output to do the same job.

If all other forms of understanding what a command does fails, at least the commands are written as pure Python functions, which all our developers can read with ease.

Please let me know if this addresses your concern, or if we can make spin better address it. We want to work with the Scientific Python community to ensure the tool serves their needs.

**Comment by rgommers** [comment]
+1 to this resolution. For NumPy I think we've not completely implemented the "command echoing" for all the custom commands, but when it's present it is very helpful for discovering all dev functionality and teaching newer contributors.

**Comment by stefanv** [comment]
@lesteve We could quite easily implement dry-run support with a decorator for the (common?) case where `run` is the only call with side-effects.

**Comment by lesteve** [comment]
Thanks for the tip about `spin.util.run` this seems very useful!

About the dry-run thing, this could be useful, but maybe `spin.util.run` is good enough, and I don't want to ask you to do any work before I look into it more.

**Comment by stefanv** [comment]
Happy to also contribute on a test branch for sklearn, just let me know.

**Comment by lagru** [comment]
I would like to add that the same arguments for making commands transparent likely applies to making environment variables that spin sets. Correct me if I'm wrong but I think spin adds to `PYTHONPATH` to run stuff in `build-install`. I realize that this might lead to more verbose commands, but I think it's equally important as command line options to understanding what's being run.

**Comment by stefanv** [comment]
Right, `spin` prints when it sets PYTHONPATH and other variables :+1:

**Comment by lagru** [comment]
Ah, I missed that. Happy with the way it already works, e.g. `export PYTHONPATH="[...]/build-install/usr/lib/python3.12/site-packages"` :+1:
